### PR TITLE
Add generate_jwt & json_{minify,prettify} helper functions

### PR DESCRIPTION
**Pull request by dwisiswant0**
## Proposed changes

This change defines several functions that manipulate JSON strings in different ways:

- The `generate_jwt` function generates a JSON Web Token (JWT) using the claims provided in a JSON string, the signature, and the specified algorithm.
- The `json_minify` function minifies a JSON string by removing unnecessary whitespace.
- The `json_prettify` function prettifies a JSON string by adding indentation.

## Checklist

- [x] Pull request is created against the [dev](https://github.com/projectdiscovery/nuclei/tree/dev) branch
- [x] All checks passed (lint, unit/integration/regression tests etc.) with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate) (https://github.com/projectdiscovery/nuclei-docs/pull/110)

### CODE REVIEW

**Comment by dwisiswant0** [comment]
## Test Case

Templates:

```yaml
// generate-jwt.yaml
id: generate-jwt-helper-functions

info:
  name: RAW Template with generate_jwt Helper Functions
  author: dwisiswant0
  severity: info

variables:
  json: |
    {
      "name": "John Doe",
      "foo": "bar"
    }
  data_compact: "{{quote_escape(json_minify(json))}}"
  data_inline: '{\"name\": \"John Doe\",\"foo\": \"bar\"}'

requests:
  - raw:
      - |+ # Escaping double quotes & minify the JSON data
        POST /generate_jwt HTTP/1.1
        Authorization: Bearer {{generate_jwt("{{data_compact}}", "hello-world", "HS256")}}

        {{data_inline}}

      - |+ # Prettify compact JSON
        POST /generate_jwt HTTP/1.1
        Authorization: Bearer {{generate_jwt("{{data_inline}}", "hello-world", "HS256")}}

        {{json_prettify("{{data_inline}}")}}

    unsafe: true
```

Usage:

```console
$ go test github.com/projectdiscovery/nuclei/v2/pkg/operators/common/...
ok      github.com/projectdiscovery/nuclei/v2/pkg/operators/common/dsl  1.029s
$ go run cmd/nuclei/main.go -duc -t generate-jwt.yaml -debug-req -u http://localhost -verbose 

                     __     _
   ____  __  _______/ /__  (_)
  / __ \/ / / / ___/ / _ \/ /
 / / / / /_/ / /__/ /  __/ /
/_/ /_/\__,_/\___/_/\___/_/   v2.8.6

        projectdiscovery.io

[INF] Using Nuclei Engine 2.8.6 (outdated)
[INF] Using Nuclei Templates  (outdated)
[INF] Templates added in last update: 0
[INF] Templates loaded for scan: 1
[INF] Targets loaded for scan: 1
[INF] [generate-jwt-helper-functions] Dumped HTTP request for http://localhost/generate_jwt

POST /generate_jwt HTTP/1.1
Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJmb28iOiJiYXIiLCJuYW1lIjoiSm9obiBEb2UifQ.EsrL8lIcYJR_Ns-JuhF3VCllCP7xwbpMCCfHin_WT6U

{\"name\": \"John Doe\",\"foo\": \"bar\"}

[INF] [generate-jwt-helper-functions] Dumped HTTP request for http://localhost/generate_jwt

POST /generate_jwt HTTP/1.1
Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJmb28iOiJiYXIiLCJuYW1lIjoiSm9obiBEb2UifQ.EsrL8lIcYJR_Ns-JuhF3VCllCP7xwbpMCCfHin_WT6U

{
    "name": "John Doe",
    "foo": "bar"
}

[INF] No results found. Better luck next time!
```

## Real Case

See https://github.com/projectdiscovery/nuclei-templates/pull/6479.

File: v2/pkg/operators/common/dsl/dsl.go [hunk 1]
```diff
@@ -809,6 +811,66 @@ func init() {
...
+			}
+
+			token := jwt.NewWithClaims(signingMethod, claims)
+
+			tokenString, err := token.SignedString(signature)
+			if err != nil {
+				return nil, err
+			}
+
+			return tokenString, nil
```
Comment (tarunKoyalwar) [inline v2/pkg/operators/common/dsl/dsl.go:841]: @dwisiswant0 , I think we are missing a attack vector here (i.e None algorithm attack).
This is a popular attack vector where some jwt vendors do not verify if algorithm is `noNe` which allows to create valid jwt.

I propose that  if algorithm is `none,None` or any mixedcase variations we can create a jwt without signature using given details (i.e base64 encode header and body). 

### Implementation
we can check if given algorithm is supported if not convert it to lowercase and trim and check if it none .
if yes using original algorithm value create jwt without signature

**Review by tarunKoyalwar — CHANGES_REQUESTED** [review]

File: v2/pkg/operators/common/dsl/dsl.go (hunk 1 above)
Comment (dwisiswant0) [reply v2/pkg/operators/common/dsl/dsl.go:841]: @tarunKoyalwar - Initially, I created the function with an _optionalAlgorithm_ input, which if not provided will use `alg="none"` to generate the token. However, it seems that it is not explained in the lib docs how to use it (track: https://github.com/golang-jwt/jwt/issues/264). While waiting for an answer or looking for an alternative way for this case, I will mark this PR as a draft - therefore I'll revert the input algorithm not to be required but optional.

**Comment by dwisiswant0** [comment]
## Test Case

Template:

```yaml
# template.yaml
id: generate-jwt-helper-functions

info:
  name: RAW Template with generate_jwt Helper Functions
  author: dwisiswant0
  severity: info

variables:
  json: |
    {
      "name": "John Doe",
      "foo": "bar"
    }
  data_compact: "{{quote_escape(json_minify(json))}}"
  data_inline: '{\"name\": \"John Doe\",\"foo\": \"bar\"}'
  maxAge: '{{to_unix_time("2034-12-30T16:30:10+00:00")}}'

requests:
  - raw:
      - |+ # Escaping double quotes & minify the JSON data
        POST /generate_jwt HTTP/1.1
        Authorization: Bearer {{generate_jwt("{{data_compact}}", "HS256", "hello-world")}}

        {{data_inline}}

      - |+ # Prettify compact JSON
        POST /generate_jwt HTTP/1.1
        Authorization: Bearer {{generate_jwt("{{data_inline}}", "HS256", "hello-world")}}

        {{json_prettify("{{data_inline}}")}}

      - |+ # With none algorithm
        GET /generate_jwt-none HTTP/1.1
        Authorization: Bearer {{generate_jwt("{{data_inline}}", "nOnE")}}

      - |+ # With empty signature
        GET /generate_jwt-empty-signature HTTP/1.1
        Authorization: Bearer {{generate_jwt("{{data_inline}}", "HS256", "")}}

      - |+ # With max age defined
        GET /generate_jwt-max-age HTTP/1.1
        Authorization: Bearer {{generate_jwt("{{data_inline}}", "HS256", "hello-world", maxAge)}}

    unsafe: true
```

Usage:

```console
$ go test github.com/projectdiscovery/nuclei/v2/pkg/operators/common/...
ok      github.com/projectdiscovery/nuclei/v2/pkg/operators/common/dsl  1.022s
$ go run cmd/nuclei/main.go -duc -t template.yaml -debug-req -u http://localhost 

                     __     _
   ____  __  _______/ /__  (_)
  / __ \/ / / / ___/ / _ \/ /
 / / / / /_/ / /__/ /  __/ /
/_/ /_/\__,_/\___/_/\___/_/   v2.8.6

        projectdiscovery.io

[INF] Using Nuclei Engine 2.8.6 (outdated)
[INF] Using Nuclei Templates  (outdated)
[INF] Templates added in last update: 0
[INF] Templates loaded for scan: 1
[INF] Targets loaded for scan: 1
[INF] [generate-jwt-helper-functions] Dumped HTTP request for http://localhost/generate_jwt

POST /generate_jwt HTTP/1.1
Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJmb28iOiJiYXIiLCJuYW1lIjoiSm9obiBEb2UifQ.EsrL8lIcYJR_Ns-JuhF3VCllCP7xwbpMCCfHin_WT6U

{\"name\": \"John Doe\",\"foo\": \"bar\"}

[INF] [generate-jwt-helper-functions] Dumped HTTP request for http://localhost/generate_jwt

POST /generate_jwt HTTP/1.1
Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJmb28iOiJiYXIiLCJuYW1lIjoiSm9obiBEb2UifQ.EsrL8lIcYJR_Ns-JuhF3VCllCP7xwbpMCCfHin_WT6U

{
    "name": "John Doe",
    "foo": "bar"
}

[INF] [generate-jwt-helper-functions] Dumped HTTP request for http://localhost/generate_jwt-none

GET /generate_jwt-none HTTP/1.1
Authorization: Bearer eyJhbGciOiJOT05FIiwidHlwIjoiSldUIn0.eyJmb28iOiJiYXIiLCJuYW1lIjoiSm9obiBEb2UifQ.

[INF] [generate-jwt-helper-functions] Dumped HTTP request for http://localhost/generate_jwt-empty-signature

GET /generate_jwt-empty-signature HTTP/1.1
Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJmb28iOiJiYXIiLCJuYW1lIjoiSm9obiBEb2UifQ.mba3WxAJP0ExW7yzDTAG3jslEQj5HfhBos6bD0R6ABQ

[INF] [generate-jwt-helper-functions] Dumped HTTP request for http://localhost/generate_jwt-max-age

GET /generate_jwt-max-age HTTP/1.1
Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJmb28iOiJiYXIiLCJuYW1lIjoiSm9obiBEb2UiLCJpYXQiOjE2NzMwNzU0OTIsImV4cCI6MjA1MTEwOTAxMH0.qWU_QpX_pezUBQJ0FBqS6iSBTAAOFILuHJk_CNdF02o

[INF] No results found. Better luck next time!
```

**Comment by dwisiswant0** [comment]
Because the input from `generate_jwt` uses an _interface{}_ type, therefore using the `quote_escape` function is not really necessary, so I decided to delete it.

Here is an update test case:

- Template:

```yaml
# template.yaml
id: generate-jwt-helper-functions

info:
  name: RAW Template with generate_jwt Helper Functions
  author: dwisiswant0
  severity: info

variables:
  json: |
    {
      "name": "John Doe",
      "foo": "bar"
    }
  json_compact: |
    {"name": "John Doe","foo": "bar"}
  maxAge: '{{to_unix_time("2034-12-30T16:30:10+00:00")}}'

requests:
  - raw:
      - |+ # Generate JWT & minify the JSON data
        POST /generate_jwt HTTP/1.1
        Authorization: Bearer {{generate_jwt(json, "HS256", "hello-world")}}
        X-Notes: This-JWT-should-be-the-same-as-the-JSON-compact-data-in-the-request-below

        {{json_minify(json)}}

      - |+ # Generate JWT with compact data & prettify
        POST /generate_jwt-prettify HTTP/1.1
        Authorization: Bearer {{generate_jwt(json_compact, "HS256", "hello-world")}}
        X-Notes: This-JWT-should-be-the-same-as-the-JSON-compact-data-in-the-request-above

        {{json_prettify(json_compact)}}

      - |+ # With none algorithm
        GET /generate_jwt-none HTTP/1.1
        Authorization: Bearer {{generate_jwt(json, "nOnE")}}

      - |+ # With empty signature
        GET /generate_jwt-empty-signature HTTP/1.1
        Authorization: Bearer {{generate_jwt(json, "HS256", "")}}

      - |+ # With max age defined
        GET /generate_jwt-max-age HTTP/1.1
        Authorization: Bearer {{generate_jwt(json, "HS256", "hello-world", maxAge)}}

    unsafe: true
```

- Usage:

```console
$ go run cmd/nuclei/main.go -duc -t template.yaml -debug-req -u http://localhost 

                     __     _
   ____  __  _______/ /__  (_)
  / __ \/ / / / ___/ / _ \/ /
 / / / / /_/ / /__/ /  __/ /
/_/ /_/\__,_/\___/_/\___/_/   v2.8.6

		projectdiscovery.io

[INF] Using Nuclei Engine 2.8.6 (outdated)
[INF] Using Nuclei Templates  (outdated)
[INF] Templates added in last update: 0
[INF] Templates loaded for scan: 1
[INF] Targets loaded for scan: 1
[INF] [generate-jwt-helper-functions] Dumped HTTP request for http://localhost/generate_jwt

POST /generate_jwt HTTP/1.1
Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJmb28iOiJiYXIiLCJuYW1lIjoiSm9obiBEb2UifQ.EsrL8lIcYJR_Ns-JuhF3VCllCP7xwbpMCCfHin_WT6U
X-Notes: This-JWT-should-be-the-same-as-the-JSON-compact-data-in-the-request-below

{"foo":"bar","name":"John Doe"}

[INF] [generate-jwt-helper-functions] Dumped HTTP request for http://localhost/generate_jwt-prettify

POST /generate_jwt-prettify HTTP/1.1
Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJmb28iOiJiYXIiLCJuYW1lIjoiSm9obiBEb2UifQ.EsrL8lIcYJR_Ns-JuhF3VCllCP7xwbpMCCfHin_WT6U
X-Notes: This-JWT-should-be-the-same-as-the-normal-JSON-data-in-the-request-above

{
    "name": "John Doe",
    "foo": "bar"
}


[INF] [generate-jwt-helper-functions] Dumped HTTP request for http://localhost/generate_jwt-none

GET /generate_jwt-none HTTP/1.1
Authorization: Bearer eyJhbGciOiJOT05FIiwidHlwIjoiSldUIn0.eyJmb28iOiJiYXIiLCJuYW1lIjoiSm9obiBEb2UifQ.

[INF] [generate-jwt-helper-functions] Dumped HTTP request for http://localhost/generate_jwt-empty-signature

GET /generate_jwt-empty-signature HTTP/1.1
Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJmb28iOiJiYXIiLCJuYW1lIjoiSm9obiBEb2UifQ.mba3WxAJP0ExW7yzDTAG3jslEQj5HfhBos6bD0R6ABQ

[INF] [generate-jwt-helper-functions] Dumped HTTP request for http://localhost/generate_jwt-max-age

GET /generate_jwt-max-age HTTP/1.1
Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJmb28iOiJiYXIiLCJuYW1lIjoiSm9obiBEb2UiLCJpYXQiOjE2NzMwODY4OTgsImV4cCI6MjA1MTEwOTAxMH0.d_HzXqMXnKN1KXVQszwQIStAFv3nIPba1qUs-9KS3iA

[INF] No results found. Better luck next time!
```

**Review by tarunKoyalwar — APPROVED** [review]
lgtm!

**Comment by tarunKoyalwar** [comment]
# Things to Note

- Dependency https://github.com/kataras/jwt
- If given algorithm is `none,None etc` . the algorithm value in jwt string is  always`NONE`

**Comment by tarunKoyalwar** [comment]
```sh

                     __     _
   ____  __  _______/ /__  (_)
  / __ \/ / / / ___/ / _ \/ /
 / / / / /_/ / /__/ /  __/ /
/_/ /_/\__,_/\___/_/\___/_/   v2.8.6

		projectdiscovery.io

[INF] Using Nuclei Engine 2.8.6 (latest)
[INF] Using Nuclei Templates 9.3.4 (latest)
[INF] Templates added in last update: 15
[INF] Templates loaded for scan: 1
[INF] Targets loaded for scan: 1
[INF] [generate-jwt-helper-functions] Dumped HTTP request for https://scanme.sh/generate_jwt

POST /generate_jwt HTTP/1.1
Host: scanme.sh
User-Agent: Mozilla/5.0 (Windows NT 6.2; WOW64) AppleWebKit/537.36 (KHTML like Gecko) Chrome/44.0.2403.155 Safari/537.36
Connection: close
Content-Length: 33
Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJmb28iOiJiYXIiLCJuYW1lIjoiSm9obiBEb2UifQ.EsrL8lIcYJR_Ns-JuhF3VCllCP7xwbpMCCfHin_WT6U
X-Notes: This-JWT-should-be-the-same-as-the-JSON-compact-data-in-the-request-below
Accept-Encoding: gzip

{"foo":"bar","name":"John Doe"}


POST /generate_jwt-prettify HTTP/1.1
Host: scanme.sh
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_8_3) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/54.0.2866.71 Safari/537.36
Connection: close
Content-Length: 47
Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJmb28iOiJiYXIiLCJuYW1lIjoiSm9obiBEb2UifQ.EsrL8lIcYJR_Ns-JuhF3VCllCP7xwbpMCCfHin_WT6U
X-Notes: This-JWT-should-be-the-same-as-the-normal-JSON-data-in-the-request-above
Accept-Encoding: gzip

{
    "name": "John Doe",
    "foo": "bar"
}



GET /generate_jwt-none HTTP/1.1
Host: scanme.sh
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/89.0.4389.114 Safari/537.36
Connection: close
Authorization: Bearer eyJhbGciOiJOT05FIiwidHlwIjoiSldUIn0.eyJmb28iOiJiYXIiLCJuYW1lIjoiSm9obiBEb2UifQ.
Accept-Encoding: gzip



GET /generate_jwt-empty-signature HTTP/1.1
Host: scanme.sh
User-Agent: Mozilla/5.0 (Windows NT 6.3; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/41.0.2226.0 Safari/537.36
Connection: close
Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJmb28iOiJiYXIiLCJuYW1lIjoiSm9obiBEb2UifQ.mba3WxAJP0ExW7yzDTAG3jslEQj5HfhBos6bD0R6ABQ
Accept-Encoding: gzip



GET /generate_jwt-max-age HTTP/1.1
Host: scanme.sh
User-Agent: Mozilla/5.0 (Windows NT 6.4; WOW64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/41.0.2225.0 Safari/537.36
Connection: close
Authorization: Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJmb28iOiJiYXIiLCJuYW1lIjoiSm9obiBEb2UiLCJpYXQiOjE2NzM2MTAyNTUsImV4cCI6MjA1MTEwOTAxMH0.9Tc-zNnc0IpVfnnLrUDLUB2WDZODj5lz1mVqkwswdeI
Accept-Encoding: gzip
```

**Comment by dwisiswant0** [comment]
> ...
> * If given algorithm is `none,None etc` . the algorithm value in jwt string is  always`NONE`

I regret to inform you that the notes provided require correction. If the supplied `optionalAlgorithm` argument is an empty string (""), `None`/`nOnE` _(any mixed-case_) - then it's set as **NONE**, otherwise, and/or not listed in case of [R844-R869](https://github.com/projectdiscovery/nuclei/pull/3160/files#diff-249444a46c383b945b14487254a7f8c1d5780cd273b8b297eb9d1a997dacf61dR844-R869) _(after converting its value to uppercase)_, it will return an error.

**Comment by tarunKoyalwar** [comment]
@dwisiswant0  , If I am not wrong If I want alg type to be `noNe` in jwt string . It is not possible here because we use `jwt.NONE` interfact/struct . 
Ex: If we take jwt string from example 3 the jwt string is
Screenshot 2023-01-13 at 5 33 14 PM

**Comment by tarunKoyalwar** [comment]
@dwisiswant0  , I explicitly wanted to mention above case but I have also updated my earlier comment

**Comment by tarunKoyalwar** [comment]
## Minor Improvements

- none jwt alg value is not converted to uppercase anymore


```yaml
# template.yaml
id: generate-jwt-helper-functions

info:
  name: RAW Template with generate_jwt Helper Functions
  author: dwisiswant0
  severity: info

variables:
  json: |
    {
      "name": "John Doe",
      "foo": "bar"
    }
  json_compact: |
    {"name": "John Doe","foo": "bar"}
  maxAge: '{{to_unix_time("2034-12-30T16:30:10+00:00")}}'

requests:
  - raw:
      - |+ # With none algorithm
        GET /generate_jwt-none HTTP/1.1
        Authorization: Bearer {{generate_jwt(json, "nOnE")}}

      - |+ # With none algorithm
        GET /generate_jwt-none HTTP/1.1
        Authorization: Bearer {{generate_jwt(json, "nONe")}}
```

```sh
./nuclei -u https://scanme.sh -t ~/test/jwtfuzz.yaml -debug-req

                     __     _
   ____  __  _______/ /__  (_)
  / __ \/ / / / ___/ / _ \/ /
 / / / / /_/ / /__/ /  __/ /
/_/ /_/\__,_/\___/_/\___/_/   v2.8.6

		projectdiscovery.io

[INF] Using Nuclei Engine 2.8.6 (latest)
[INF] Using Nuclei Templates 9.3.4 (latest)
[INF] Templates added in last update: 15
[INF] Templates loaded for scan: 1
[INF] Targets loaded for scan: 1
[INF] [generate-jwt-helper-functions] Dumped HTTP request for https://scanme.sh/generate_jwt-none

GET /generate_jwt-none HTTP/1.1
Host: scanme.sh
User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10_10_1) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/37.0.2062.124 Safari/537.36
Connection: close
Authorization: Bearer eyJhbGciOiJuT25FIiwidHlwIjoiSldUIn0.eyJmb28iOiJiYXIiLCJuYW1lIjoiSm9obiBEb2UifQ.
Accept-Encoding: gzip

[INF] [generate-jwt-helper-functions] Dumped HTTP request for https://scanme.sh/generate_jwt-none

GET /generate_jwt-none HTTP/1.1
Host: scanme.sh
User-Agent: Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/34.0.1847.137 Safari/4E423F
Connection: close
Authorization: Bearer eyJhbGciOiJuT05lIiwidHlwIjoiSldUIn0.eyJmb28iOiJiYXIiLCJuYW1lIjoiSm9obiBEb2UifQ.
Accept-Encoding: gzip

[INF] No results found. Better luck next time!
```

**Comment by tarunKoyalwar** [comment]
Screenshot 2023-01-13 at 6 02 42 PM

**Review by ehsandeep — APPROVED** [review]

**Pull request merged**